daemon/containerd/image_changes.go

```diff
@@ -0,0 +1,98 @@
+package containerd
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/mount"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/google/uuid"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+func (i *ImageService) Changes(ctx context.Context, container *container.Container) ([]archive.Change, error) {
+	snapshotter := i.client.SnapshotService(i.snapshotter)
+	mounts, err := snapshotter.Mounts(ctx, container.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	cs := i.client.ContentStore()
+
+	imageManifestBytes, err := content.ReadBlob(ctx, cs, *container.ImageManifest)
+	if err != nil {
+		return nil, err
+	}
+
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(imageManifestBytes, &manifest); err != nil {
+		return nil, err
+	}
+
+	imageConfigBytes, err := content.ReadBlob(ctx, cs, manifest.Config)
+	if err != nil {
+		return nil, err
+	}
+	var image ocispec.Image
+	if err := json.Unmarshal(imageConfigBytes, &image); err != nil {
+		return nil, err
+	}
+
+	diffIDs := image.RootFS.DiffIDs
+	rnd, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+	parent, err := snapshotter.View(ctx, rnd.String(), identity.ChainID(diffIDs).String())
+	if err != nil {
+		return nil, err
+	}
+	defer snapshotter.Remove(ctx, rnd.String())
+
+	var changes []archive.Change
+	err = mount.WithTempMount(ctx, readOnly(mounts), func(fs string) error {
+		return mount.WithTempMount(ctx, parent, func(root string) error {
+			changes, err = archive.ChangesDirs(fs, root)
+			return err
+		})
+	})
+	return changes, err
+}
+
+func readOnly(mounts []mount.Mount) []mount.Mount {
+	for i, m := range mounts {
+		if m.Type == "overlay" {
+			opts := make([]string, 0, len(m.Options))
+			upper := ""
+			for _, o := range m.Options {
```

Comment:
This piece of code might need a comment or something. This is needed so that we can mount the image a second time but in read only. I know that Buildkit has kinda the same code somewhere, @tonistiigi even opened a PR to containerd to add this ability (I think) but was never merged.

We need this to be read only for overlayfs because if we don't the mounts are in an undefined state and you get a lot of "No found" errors, basically nothing works.

As I understand it, this is only an issue for the overlayfs driver, other drivers handle multiple mounts with no issues.

Reply:
Here's where it's done in buildkit's code: https://github.com/moby/buildkit/blob/75123c696506bdbca1ed69906479e200f1b62604/cache/refs.go#L1516

Reply:
Thanks for the extra details. Yes, I understood we needed a read-only mount here (but wasn't sure if other drivers could have their own "special needs"). We should definitely have a comment in this part of the code describing some of this, and we should try to resurrect the work in Containerd if possible.

The part that somewhat scares me with such code is that there may be other options to be used when mounting, such as

- `userxattr` and `index=on/off`; https://github.com/moby/moby/blob/ad9d70b0e6833b746eecccc08a17459ba8bb31c4/daemon/graphdriver/overlay2/overlay.go#L564
- selinux labels; https://github.com/moby/moby/blob/ad9d70b0e6833b746eecccc08a17459ba8bb31c4/daemon/graphdriver/overlay2/overlay.go#L569
- for the mount-paths, "special trickery" to make sure we can mount all layers https://github.com/moby/moby/blob/ad9d70b0e6833b746eecccc08a17459ba8bb31c4/daemon/graphdriver/overlay2/overlay.go#L580-L584

Missing some of those may be hard to miss ("it works" in most cases ... until it doesn't), which can make them hard to debug.

Reply:
I hear you, but these options you linked is the creation of the mounts, we are not creating any mounts, it's the snapshotter that does it and we are only changing the options related to the (upper/work) directories, other options are left unchanged.

I can't find the PR I saw a couple months back, I'll continue looking for it

Reply:
Found them!

* https://github.com/moby/buildkit/pull/1100
* https://github.com/containerd/containerd/pull/6114

Reply:
My instinct is that we should try and get this code into containerd before baking knowledge of the implementation of the overlay snapshotter internals into Moby (which is relatively stable as it's derived from our overlay2 graphdriver, but still is very fraught) -- is there a willingness to pick up that containerd PR and finish it?

Reply:
I just had a second look at the c8d PR and it works with the diff compare only. IIUC containerd has differs and comparers but puts the results in the content store, I'm not sure we want to do that, reusing the `docker/docker/pkg/archive` feels nice here

Reply:
> baking knowledge of the implementation of the overlay snapshotter internals

I see it more as "baking knowledge of how overlay works" then it's easier to swallow the pill

Reply:
Those are one and the same I think -- we're coupling specifically to the overlay2-style implementation and how the mounts are constructed. I don't think this is an abstraction that makes sense to leak.

Reply:
> and we are only changing the options related to the (upper/work) directories, other options are left unchanged.

Good point, guess I overlooked that (and read too much into it)

> My instinct is that we should try and get this code into containerd before baking knowledge of the implementation of the overlay snapshotter internals into Moby

Yes, I agree, and what led me to my comment.

With the added information (scope is fairly limited, and BuildKit already used this and the approach is somewhat battle-tested), I'm good with taking it (for now) while we work on looking if we can upstream these things.

My suggestion is;

- Create a tracking ticket (or link to an existing one if we have)
- Add a TODO comment (or at least some comment describing the "why", and "what we would like to see in future")
- Add a comment with a link to the corresponding BuildKit code (this can help finding the implementation there, and to verify we're still "in sync" with eachother, which can be a big time-saver)


With that in place, I'm good with getting this change in (as "good", not "perfect").

Reply:
Complete ignorance on this end; if there's an API designed for this use case we should probably just use that :laughing:

Reply:
Reading the doc and it might not do what we want, we should take a quick look though. There _must_ be a reason buildkit does it this way and not with a `View`

Reply:
I don't think we're "stuck" -- that PR appears to have stalled in review because the contributor lost interest. I think anyone interested could pick it up and get it past the review hurdles; now is probably better than later as well as containerd 1.7 is about to be cut in the next couple weeks.

Reply:
Looking back at @thaJeztah's comments, I wonder – a lot of the code between `docker diff` and `docker commit` should be the same. I wonder if we can refactor some of that and share it, we're already using `rootfs.CreateDiff()` in the `docker commit` code, I think this could be easier and a bit more streamlined and we could reuse some code here.

Reply:
Looks like `snapshotter.View` doesn't really do what we want here – we use `snapshotter.Mounts` to get the active snapshot we've previously created for the container in `PrepareSnapshot` w/ `snapshotter.Prepare`. `snapshotter.View` just gets a new readonly view for the parent, which isn't what we want in `docker diff`.

I'll try to get that code merged into containerd, and in the meantime leave this as is with a todo to use the containerd code when it gets merged, if that's alright.

Reply:
https://github.com/containerd/containerd/pull/8259